core/src/main/java/com/linecorp/armeria/common/HttpRequest.java

```diff
@@ -525,4 +528,41 @@ default HttpRequestDuplicator toDuplicator(EventExecutor executor, long maxReque
                                         Function<? super HttpData, ? extends ByteBuf> byteBufConverter) {
         return new DecodedHttpStreamMessage<>(this, decoder, alloc, byteBufConverter);
     }
+
+    /**
+     * Transforms the {@link ResponseHeaders} of this {@link HttpRequest} by applying the specified
+     * {@link Function}.
+     */
+    default HttpRequest mapHeaders(Function<? super RequestHeaders, ? extends RequestHeaders> function) {
+        requireNonNull(function, "function");
+        final RequestHeaders transformed = function.apply(headers());
+        requireNonNull(transformed, "function.apply() returned null");
+        return withHeaders(transformed);
+    }
+
+    /**
+     * Transforms the {@link HttpData}s emitted by this {@link HttpRequest} by applying the specified
+     * {@link Function}.
+     */
+    default HttpRequest mapData(Function<? super HttpData, ? extends HttpData> function) {
+        requireNonNull(function, "function");
+        final StreamMessage<HttpObject> stream =
+                map(obj -> obj instanceof HttpData ? function.apply((HttpData) obj) : obj);
+        return of(headers(), stream);
+    }
+
+    /**
+     * Transforms the {@linkplain HttpHeaders trailers} emitted by this {@link HttpRequest} by applying the
+     * specified {@link Function}.
+     */
+    default HttpRequest mapTrailers(Function<? super HttpHeaders, ? extends HttpHeaders> function) {
+        requireNonNull(function, "function");
+        final StreamMessage<HttpObject> stream = map(obj -> {
+            if (obj instanceof HttpHeaders && !(obj instanceof RequestHeaders)) {
```

Comment:
We probably don't need `!(obj instanceof RequestHeaders)`?

Reply:
Ah... nice advice!